File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rev(self, node):
        if not node or not node.next:
            return node
        prev = self.rev(node.next)
        node.next.next = node
        node.next = None
        return prev
    
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if not head or not head.next:
            return
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        head2 = self.rev(slow.next)
        slow.next = None

        
        # merge
        head1 = head
        
        while head2:
            temp = head1.next
            head1.next = head2
            head1 = head2
            head2 = temp
        
        
        

File: 0143-reorder-list/0143-reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseLinkedList(ListNode head) {
        if (head == null || head.next == null) return head;
        
        ListNode prev = reverseLinkedList(head.next);
        head.next.next = head;
        head.next = null;
        return prev;
    }
    public void reorderList(ListNode head) {
        if (head == null || head.next == null) return;
        ListNode fast = head;
        ListNode slow = head;
        
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }
        ListNode head2 = reverseLinkedList(slow.next);
        slow.next = null;
        ListNode head1 = head;
        ListNode tempNode;
        while (head2 != null) {
            tempNode = head1.next;
            head1.next = head2;
            head1 = head2;
            head2 = tempNode;
            
        }
        
        
        
    }
}